# Proyek Analisis Data: E-Commerce Public Dataset 
- **Nama:** Mochamad Nurul Huda
- **Email:** mochamadnurulhuda16@gmail.com
- **ID Dicoding:** huda16

## Menentukan Pertanyaan Bisnis

- Pertanyaan 1
- Pertanyaan 2

## Import Semua Packages/Library yang Digunakan

In [73]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import urllib
import matplotlib.image as mpimg
from scipy import stats

## Data Wrangling

### Gathering Data

In [74]:
customers_df = pd.read_csv("./dataset/customers_dataset.csv")
customers_df.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

In [75]:
geolocation_df = pd.read_csv("./dataset/geolocation_dataset.csv")
geolocation_df.head()

geolocation_df_zip_code_prefix  geolocation_df_lat  geolocation_df_lng  \
0                            1037          -23.545621          -46.639292   
1                            1046          -23.546081          -46.644820   
2                            1046          -23.546129          -46.642951   
3                            1041          -23.544392          -46.639499   
4                            1035          -23.541578          -46.641607   

  geolocation_df_city geolocation_df_state  
0           sao paulo                   SP  
1           sao paulo                   SP  
2           sao paulo                   SP  
3           sao paulo                   SP  
4           sao paulo                   SP

In [76]:
items_df = pd.read_csv("./dataset/order_items_dataset.csv")
items_df.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

In [77]:
payments_df = pd.read_csv("./dataset/order_payments_dataset.csv")
payments_df.head()

order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45

In [78]:
reviews_df = pd.read_csv("./dataset/order_reviews_dataset.csv")
reviews_df.head()

review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   

   review_score review_comment_title  \
0             4                  NaN   
1             5                  NaN   
2             5                  NaN   
3             5                  NaN   
4             5                  NaN   

                              review_comment_message review_creation_date  \
0                                                NaN  2018-01-18 00:00:00   
1                                                NaN  2018-03-10 00:00:00   
2                                                NaN  2018-02-17 00:00:00   
3              Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4  Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   

  review_answer_timestamp  
0     2018-01-18 21:46:59  
1     2018-03-11 03:05:13  
2     2018-02-18 14:36:24  
3     2017-04-21 22:02:06  
4     2018-03-02 10:26:53

In [79]:
orders_df = pd.read_csv("./dataset/orders_dataset.csv")
orders_df.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

In [80]:
category_df = pd.read_csv("./dataset/product_category_name_translation.csv")
category_df.head()

product_category_df_name product_category_df_name_english
0             beleza_saude                    health_beauty
1   informatica_acessorios            computers_accessories
2               automotivo                             auto
3          cama_mesa_banho                   bed_bath_table
4         moveis_decoracao                  furniture_decor

In [81]:
products_df = pd.read_csv("./dataset/products_dataset.csv")
products_df.head()

product_id product_category_df_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5               perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                    artes   
2  96bd76ec8810374ed1b65e291975717f            esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                    bebes   
4  9dc1a7de274444849c219cff195d0b71    utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

In [82]:

sellers_df = pd.read_csv("./dataset/sellers_dataset.csv")
sellers_df.head()

seller_id  seller_zip_code_prefix  \
0  3442f8959a84dea7ee197c632cb2df15                   13023   
1  d1b65fc7debc3361ea86b5f14c68d2e2                   13844   
2  ce3ad9de960102d0677a81f5d0bb7b2d                   20031   
3  c0f3eea2e14555b6faeea3dd58c1b1c3                    4195   
4  51a04a8a6bdcb23deccc82b0b80742cf                   12914   

         seller_city seller_state  
0           campinas           SP  
1         mogi guacu           SP  
2     rio de janeiro           RJ  
3          sao paulo           SP  
4  braganca paulista           SP

In [83]:
# Distribusi Kategori Produk
category_df_distribution = products_df['product_category_df_name'].value_counts()
print("Distribusi Kategori Produk:")
print(category_df_distribution)

# Distribusi Pelanggan Berdasarkan Kota
customer_city_distribution = customers_df['customer_city'].value_counts().head(10)
print("\nDistribusi Pelanggan Berdasarkan Kota (Top 10):")
print(customer_city_distribution)

# Distribusi Penjual Berdasarkan Negara Bagian
seller_state_distribution = sellers_df['seller_state'].value_counts()
print("\nDistribusi Penjual Berdasarkan Negara Bagian:")
print(seller_state_distribution)

# Distribusi Skor Ulasan
review_score_distribution = reviews_df['review_score'].value_counts()
print("\nDistribusi Skor Ulasan:")
print(review_score_distribution)

Distribusi Kategori Produk:
product_category_df_name
cama_mesa_banho                  3029
esporte_lazer                    2867
moveis_decoracao                 2657
beleza_saude                     2444
utilidades_domesticas            2335
                                 ... 
fashion_roupa_infanto_juvenil       5
casa_conforto_2                     5
pc_gamer                            3
seguros_e_servicos                  2
cds_dvds_musicais                   1
Name: count, Length: 73, dtype: int64

Distribusi Pelanggan Berdasarkan Kota (Top 10):
customer_city
sao paulo                15540
rio de janeiro            6882
belo horizonte            2773
brasilia                  2131
curitiba                  1521
campinas                  1444
porto alegre              1379
salvador                  1245
guarulhos                 1189
sao bernardo do campo      938
Name: count, dtype: int64

Distribusi Penjual Berdasarkan Negara Bagian:
seller_state
SP    1849
PR     349
MG     244

**Insight:**
1. **Distribusi Kategori Produk:**
    - Kategori produk dengan jumlah terbanyak adalah `cama_mesa_banho` dengan 3029 produk, diikuti oleh `esporte_lazer` dengan 2867 produk, dan `moveis_decoracao` dengan 2657 produk.
    - Kategori dengan jumlah produk paling sedikit adalah `cds_dvds_musicais` dengan hanya 1 produk, diikuti oleh `seguros_e_servicos` dengan 2 produk, dan `pc_gamer` dengan 3 produk.

2. **Distribusi Pelanggan Berdasarkan Kota:**
    - Kota dengan jumlah pelanggan terbanyak adalah `sao paulo` dengan 15540 pelanggan, diikuti oleh `rio de janeiro` dengan 6882 pelanggan, dan `belo horizonte` dengan 2773 pelanggan.
    - Kota dengan jumlah pelanggan paling sedikit di antara 10 besar adalah `sao bernardo do campo` dengan 938 pelanggan.

3. **Distribusi Penjual Berdasarkan Negara Bagian:**
    - Negara bagian dengan jumlah penjual terbanyak adalah `SP` (São Paulo) dengan 1849 penjual, diikuti oleh `PR` (Paraná) dengan 349 penjual, dan `MG` (Minas Gerais) dengan 244 penjual.
    - Negara bagian dengan jumlah penjual paling sedikit adalah `PA` (Pará), `AM` (Amazonas), `MA` (Maranhão), `AC` (Acre), dan `PI` (Piauí) masing-masing dengan 1 penjual.

4. **Distribusi Skor Ulasan:**
    - Skor ulasan terbanyak adalah 5 dengan 57328 ulasan, diikuti oleh skor 4 dengan 19142 ulasan, dan skor 1 dengan 11424 ulasan.
    - Skor ulasan paling sedikit adalah 2 dengan 3151 ulasan.

### Assessing Data

#### Menilai tabel `customers_df`

In [84]:
customers_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB


In [85]:
customers_df.isna().sum()

customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

In [86]:
customers_df.nunique()

customer_id                 99441
customer_unique_id          96096
customer_zip_code_prefix    14994
customer_city                4119
customer_state                 27
dtype: int64

In [87]:
print("Jumlah duplikasi: ", customers_df.duplicated().sum())

Jumlah duplikasi:  0


In [88]:
customers_df.describe()

customer_zip_code_prefix
count              99441.000000
mean               35137.474583
std                29797.938996
min                 1003.000000
25%                11347.000000
50%                24416.000000
75%                58900.000000
max                99990.000000

#### Menilai tabel `geolocation_df`

In [89]:
geolocation_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 #   Column                          Non-Null Count    Dtype  
---  ------                          --------------    -----  
 0   geolocation_df_zip_code_prefix  1000163 non-null  int64  
 1   geolocation_df_lat              1000163 non-null  float64
 2   geolocation_df_lng              1000163 non-null  float64
 3   geolocation_df_city             1000163 non-null  object 
 4   geolocation_df_state            1000163 non-null  object 
dtypes: float64(2), int64(1), object(2)
memory usage: 38.2+ MB


In [90]:
geolocation_df.isna().sum()

geolocation_df_zip_code_prefix    0
geolocation_df_lat                0
geolocation_df_lng                0
geolocation_df_city               0
geolocation_df_state              0
dtype: int64

In [91]:
geolocation_df.nunique()

geolocation_df_zip_code_prefix     19015
geolocation_df_lat                717360
geolocation_df_lng                717613
geolocation_df_city                 8011
geolocation_df_state                  27
dtype: int64

In [92]:
print("Jumlah duplikasi: ", geolocation_df.duplicated().sum())

Jumlah duplikasi:  261831


In [93]:
geolocation_df.describe()

geolocation_df_zip_code_prefix  geolocation_df_lat  geolocation_df_lng
count                    1.000163e+06        1.000163e+06        1.000163e+06
mean                     3.657417e+04       -2.117615e+01       -4.639054e+01
std                      3.054934e+04        5.715866e+00        4.269748e+00
min                      1.001000e+03       -3.660537e+01       -1.014668e+02
25%                      1.107500e+04       -2.360355e+01       -4.857317e+01
50%                      2.653000e+04       -2.291938e+01       -4.663788e+01
75%                      6.350400e+04       -1.997962e+01       -4.376771e+01
max                      9.999000e+04        4.506593e+01        1.211054e+02

#### Menilai tabel `items_df`

In [94]:
items_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB


In [95]:
items_df.isna().sum()

order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

In [96]:
items_df.nunique()

order_id               98666
order_item_id             21
product_id             32951
seller_id               3095
shipping_limit_date    93318
price                   5968
freight_value           6999
dtype: int64

In [97]:
print("Jumlah duplikasi: ", customers_df.duplicated().sum())

Jumlah duplikasi:  0


In [98]:
items_df.describe()

order_item_id          price  freight_value
count  112650.000000  112650.000000  112650.000000
mean        1.197834     120.653739      19.990320
std         0.705124     183.633928      15.806405
min         1.000000       0.850000       0.000000
25%         1.000000      39.900000      13.080000
50%         1.000000      74.990000      16.260000
75%         1.000000     134.900000      21.150000
max        21.000000    6735.000000     409.680000

#### Menilai tabel `payments_df`

In [99]:
payments_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  object 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  object 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.0+ MB


In [100]:
payments_df.isna().sum()

order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64

In [101]:
payments_df.nunique()

order_id                99440
payment_sequential         29
payment_type                5
payment_installments       24
payment_value           29077
dtype: int64

In [102]:
print("Jumlah duplikasi: ", payments_df.duplicated().sum())

Jumlah duplikasi:  0


In [103]:
payments_df.describe()

payment_sequential  payment_installments  payment_value
count       103886.000000         103886.000000  103886.000000
mean             1.092679              2.853349     154.100380
std              0.706584              2.687051     217.494064
min              1.000000              0.000000       0.000000
25%              1.000000              1.000000      56.790000
50%              1.000000              1.000000     100.000000
75%              1.000000              4.000000     171.837500
max             29.000000             24.000000   13664.080000

#### Menilai tabel `reviews_df`

In [104]:
reviews_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   review_id                99224 non-null  object
 1   order_id                 99224 non-null  object
 2   review_score             99224 non-null  int64 
 3   review_comment_title     11568 non-null  object
 4   review_comment_message   40977 non-null  object
 5   review_creation_date     99224 non-null  object
 6   review_answer_timestamp  99224 non-null  object
dtypes: int64(1), object(6)
memory usage: 5.3+ MB


In [105]:
reviews_df.isna().sum()

review_id                      0
order_id                       0
review_score                   0
review_comment_title       87656
review_comment_message     58247
review_creation_date           0
review_answer_timestamp        0
dtype: int64

In [106]:
reviews_df.nunique()

review_id                  98410
order_id                   98673
review_score                   5
review_comment_title        4527
review_comment_message     36159
review_creation_date         636
review_answer_timestamp    98248
dtype: int64

In [107]:
print("Jumlah duplikasi: ", reviews_df.duplicated().sum())

Jumlah duplikasi:  0


In [108]:
reviews_df.describe()

review_score
count  99224.000000
mean       4.086421
std        1.347579
min        1.000000
25%        4.000000
50%        5.000000
75%        5.000000
max        5.000000

#### Menilai tabel `orders_df`

In [109]:
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB


In [110]:
orders_df.isna().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

In [111]:
orders_df.nunique()

order_id                         99441
customer_id                      99441
order_status                         8
order_purchase_timestamp         98875
order_approved_at                90733
order_delivered_carrier_date     81018
order_delivered_customer_date    95664
order_estimated_delivery_date      459
dtype: int64

In [112]:
print("Jumlah duplikasi: ", orders_df.duplicated().sum())

Jumlah duplikasi:  0


In [113]:
orders_df.describe()

order_id                       customer_id  \
count                              99441                             99441   
unique                             99441                             99441   
top     e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
freq                                   1                                 1   

       order_status order_purchase_timestamp    order_approved_at  \
count         99441                    99441                99281   
unique            8                    98875                90733   
top       delivered      2018-04-11 10:48:14  2018-02-27 04:31:10   
freq          96478                        3                    9   

       order_delivered_carrier_date order_delivered_customer_date  \
count                         97658                         96476   
unique                        81018                         95664   
top             2018-05-09 15:48:00           2018-05-08 23:38:46   
freq                             47                             3   

       order_estimated_delivery_date  
count                          99441  
unique                           459  
top              2017-12-20 00:00:00  
freq                             522

#### Menilai tabel `category_df`

In [114]:
category_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71 entries, 0 to 70
Data columns (total 2 columns):
 #   Column                            Non-Null Count  Dtype 
---  ------                            --------------  ----- 
 0   product_category_df_name          71 non-null     object
 1   product_category_df_name_english  71 non-null     object
dtypes: object(2)
memory usage: 1.2+ KB


In [115]:
category_df.isna().sum()

product_category_df_name            0
product_category_df_name_english    0
dtype: int64

In [116]:
category_df.nunique()

product_category_df_name            71
product_category_df_name_english    71
dtype: int64

In [117]:
print("Jumlah duplikasi: ", category_df.duplicated().sum())

Jumlah duplikasi:  0


In [118]:
category_df.describe()

product_category_df_name product_category_df_name_english
count                        71                               71
unique                       71                               71
top                beleza_saude                    health_beauty
freq                          1                                1

#### Menilai tabel `products_df`

In [119]:
products_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_df_name    32341 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB


In [120]:
products_df.isna().sum()

product_id                      0
product_category_df_name      610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

In [121]:
products_df.nunique()

product_id                    32951
product_category_df_name         73
product_name_lenght              66
product_description_lenght     2960
product_photos_qty               19
product_weight_g               2204
product_length_cm                99
product_height_cm               102
product_width_cm                 95
dtype: int64

In [122]:
print("Jumlah duplikasi: ", products_df.duplicated().sum())

Jumlah duplikasi:  0


In [123]:
products_df.describe()

product_name_lenght  product_description_lenght  product_photos_qty  \
count         32341.000000                32341.000000        32341.000000   
mean             48.476949                  771.495285            2.188986   
std              10.245741                  635.115225            1.736766   
min               5.000000                    4.000000            1.000000   
25%              42.000000                  339.000000            1.000000   
50%              51.000000                  595.000000            1.000000   
75%              57.000000                  972.000000            3.000000   
max              76.000000                 3992.000000           20.000000   

       product_weight_g  product_length_cm  product_height_cm  \
count      32949.000000       32949.000000       32949.000000   
mean        2276.472488          30.815078          16.937661   
std         4282.038731          16.914458          13.637554   
min            0.000000           7.000000           2.000000   
25%          300.000000          18.000000           8.000000   
50%          700.000000          25.000000          13.000000   
75%         1900.000000          38.000000          21.000000   
max        40425.000000         105.000000         105.000000   

       product_width_cm  
count      32949.000000  
mean          23.196728  
std           12.079047  
min            6.000000  
25%           15.000000  
50%           20.000000  
75%           30.000000  
max          118.000000

#### Menilai tabel `sellers_df`

In [124]:
sellers_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   seller_id               3095 non-null   object
 1   seller_zip_code_prefix  3095 non-null   int64 
 2   seller_city             3095 non-null   object
 3   seller_state            3095 non-null   object
dtypes: int64(1), object(3)
memory usage: 96.8+ KB


In [125]:
sellers_df.isna().sum()

seller_id                 0
seller_zip_code_prefix    0
seller_city               0
seller_state              0
dtype: int64

In [126]:
sellers_df.nunique()

seller_id                 3095
seller_zip_code_prefix    2246
seller_city                611
seller_state                23
dtype: int64

In [127]:
print("Jumlah duplikasi: ", sellers_df.duplicated().sum())

Jumlah duplikasi:  0


In [128]:
sellers_df.describe()

seller_zip_code_prefix
count             3095.000000
mean             32291.059451
std              32713.453830
min               1001.000000
25%               7093.500000
50%              14940.000000
75%              64552.500000
max              99730.000000

**Insight:**
1. **Tabel `customers`:**
    - Tidak ada nilai yang hilang pada tabel `customers`.
    - Terdapat 99441 entri unik pada tabel ini.
    - Tidak ada duplikasi data pada tabel `customers`.
    - Kolom `customer_zip_code_prefix` memiliki tipe data `int64`, sedangkan kolom lainnya memiliki tipe data `object`.

2. **Tabel `geolocation`:**
    - Tidak ada nilai yang hilang pada tabel `geolocation`.
    - Terdapat 1000163 entri unik pada tabel ini.
    - Tidak ada duplikasi data pada tabel `geolocation`.
    - Kolom `geolocation_lat` dan `geolocation_lng` memiliki tipe data `float64`, sedangkan kolom lainnya memiliki tipe data `int64` dan `object`.

3. **Tabel `items`:**
    - Tidak ada nilai yang hilang pada tabel `items`.
    - Terdapat 112650 entri unik pada tabel ini.
    - Tidak ada duplikasi data pada tabel `items`.
    - Kolom `price` dan `freight_value` memiliki tipe data `float64`, kolom `order_item_id` memiliki tipe data `int64`, sedangkan kolom lainnya memiliki tipe data `object`.

4. **Tabel `payments`:**
    - Tidak ada nilai yang hilang pada tabel `payments`.
    - Terdapat 103886 entri unik pada tabel ini.
    - Tidak ada duplikasi data pada tabel `payments`.
    - Kolom `payment_value` memiliki tipe data `float64`, kolom `payment_sequential` dan `payment_installments` memiliki tipe data `int64`, sedangkan kolom lainnya memiliki tipe data `object`.

5. **Tabel `reviews`:**
    - Terdapat beberapa nilai yang hilang pada kolom `review_comment_title` dan `review_comment_message`.
    - Terdapat 99224 entri unik pada tabel ini.
    - Tidak ada duplikasi data pada tabel `reviews`.
    - Kolom `review_score` memiliki tipe data `int64`, sedangkan kolom lainnya memiliki tipe data `object`.

6. **Tabel `orders`:**
    - Terdapat beberapa nilai yang hilang pada kolom `order_approved_at`, `order_delivered_carrier_date`, dan `order_delivered_customer_date`.
    - Terdapat 99441 entri unik pada tabel ini.
    - Tidak ada duplikasi data pada tabel `orders`.
    - Semua kolom pada tabel ini memiliki tipe data `object`.

7. **Tabel `category`:**
    - Tidak ada nilai yang hilang pada tabel `category`.
    - Terdapat 71 entri unik pada tabel ini.
    - Tidak ada duplikasi data pada tabel `category`.
    - Semua kolom pada tabel ini memiliki tipe data `object`.

8. **Tabel `products`:**
    - Terdapat beberapa nilai yang hilang pada kolom `product_category_name`, `product_name_lenght`, `product_description_lenght`, `product_photos_qty`, `product_weight_g`, `product_length_cm`, `product_height_cm`, dan `product_width_cm`.
    - Terdapat 32951 entri unik pada tabel ini.
    - Tidak ada duplikasi data pada tabel `products`.
    - Kolom `product_weight_g`, `product_length_cm`, `product_height_cm`, `product_width_cm`, `product_name_lenght`, `product_description_lenght`, dan `product_photos_qty` memiliki tipe data `float64`, sedangkan kolom lainnya memiliki tipe data `object`.

9. **Tabel `sellers`:**
    - Tidak ada nilai yang hilang pada tabel `sellers`.
    - Terdapat 3095 entri unik pada tabel ini.
    - Tidak ada duplikasi data pada tabel `sellers`.
    - Kolom `seller_zip_code_prefix` memiliki tipe data `int64`, sedangkan kolom lainnya memiliki tipe data `object`.

### Cleaning Data

##### Duplicate data

In [129]:
# Remove duplicates from customers_df DataFrame
customers_df.drop_duplicates(inplace=True)
print("Jumlah duplikasi: ", customers_df.duplicated().sum())

# Remove duplicates from geolocation_df DataFrame
geolocation_df.drop_duplicates(inplace=True)
print("Jumlah duplikasi: ", geolocation_df.duplicated().sum())

# Remove duplicates from items_df DataFrame
items_df.drop_duplicates(inplace=True)
print("Jumlah duplikasi: ", items_df.duplicated().sum())

# Remove duplicates from payments_df DataFrame
payments_df.drop_duplicates(inplace=True)
print("Jumlah duplikasi: ", payments_df.duplicated().sum())

# Remove duplicates from reviews_df DataFrame
reviews_df.drop_duplicates(inplace=True)
print("Jumlah duplikasi: ", reviews_df.duplicated().sum())

# Remove duplicates from orders_df DataFrame
orders_df.drop_duplicates(inplace=True)
print("Jumlah duplikasi: ", orders_df.duplicated().sum())

# Remove duplicates from category_df DataFrame
category_df.drop_duplicates(inplace=True)
print("Jumlah duplikasi: ", category_df.duplicated().sum())

# Remove duplicates from products_df DataFrame
products_df.drop_duplicates(inplace=True)
print("Jumlah duplikasi: ", products_df.duplicated().sum())

# Remove duplicates from sellers_df DataFrame
sellers_df.drop_duplicates(inplace=True)
print("Jumlah duplikasi: ", sellers_df.duplicated().sum())

Jumlah duplikasi:  0
Jumlah duplikasi:  0
Jumlah duplikasi:  0
Jumlah duplikasi:  0
Jumlah duplikasi:  0
Jumlah duplikasi:  0
Jumlah duplikasi:  0
Jumlah duplikasi:  0
Jumlah duplikasi:  0


##### Missing Value

In [130]:
# Handling missing values in the 'products_df' DataFrame
products_df['product_category_df_name'].fillna('unknown', inplace=True)
products_df['product_name_lenght'].fillna(products_df['product_name_lenght'].mean(), inplace=True)
products_df['product_description_lenght'].fillna(products_df['product_description_lenght'].mean(), inplace=True)
products_df['product_photos_qty'].fillna(products_df['product_photos_qty'].mean(), inplace=True)
products_df['product_weight_g'].fillna(products_df['product_weight_g'].mean(), inplace=True)
products_df['product_length_cm'].fillna(products_df['product_length_cm'].mean(), inplace=True)
products_df['product_height_cm'].fillna(products_df['product_height_cm'].mean(), inplace=True)
products_df['product_width_cm'].fillna(products_df['product_width_cm'].mean(), inplace=True)

# Handling missing values in the 'orders_df' DataFrame
orders_df['order_approved_at'].fillna(method='ffill', inplace=True)
orders_df['order_delivered_carrier_date'].fillna(method='ffill', inplace=True)
orders_df['order_delivered_customer_date'].fillna(method='ffill', inplace=True)

# Handling missing values in the 'reviews_df' DataFrame
reviews_df['review_comment_title'].fillna('No Title', inplace=True)
reviews_df['review_comment_message'].fillna('No Comment', inplace=True)

# Verify that there are no missing values left
print(products_df.isna().sum())
print(orders_df.isna().sum())
print(reviews_df.isna().sum())

product_id                    0
product_category_df_name      0
product_name_lenght           0
product_description_lenght    0
product_photos_qty            0
product_weight_g              0
product_length_cm             0
product_height_cm             0
product_width_cm              0
dtype: int64
order_id                         0
customer_id                      0
order_status                     0
order_purchase_timestamp         0
order_approved_at                0
order_delivered_carrier_date     0
order_delivered_customer_date    0
order_estimated_delivery_date    0
dtype: int64
review_id                  0
order_id                   0
review_score               0
review_comment_title       0
review_comment_message     0
review_creation_date       0
review_answer_timestamp    0
dtype: int64


C:\Users\mocha\AppData\Local\Temp\ipykernel_508\1791806087.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  products_df['product_category_df_name'].fillna('unknown', inplace=True)
C:\Users\mocha\AppData\Local\Temp\ipykernel_508\1791806087.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves 

##### Inaccurate Value

In [131]:
# Handling inaccurate values in the 'products_df' DataFrame
# Removing outliers in 'product_weight_g'
products_df = products_df[products_df['product_weight_g'] < products_df['product_weight_g'].quantile(0.99)]

# Removing outliers in 'product_length_cm', 'product_height_cm', and 'product_width_cm'
products_df = products_df[products_df['product_length_cm'] < products_df['product_length_cm'].quantile(0.99)]
products_df = products_df[products_df['product_height_cm'] < products_df['product_height_cm'].quantile(0.99)]
products_df = products_df[products_df['product_width_cm'] < products_df['product_width_cm'].quantile(0.99)]

# Correcting data types if necessary
orders_df['order_purchase_timestamp'] = pd.to_datetime(orders_df['order_purchase_timestamp'])
orders_df['order_approved_at'] = pd.to_datetime(orders_df['order_approved_at'])
orders_df['order_delivered_carrier_date'] = pd.to_datetime(orders_df['order_delivered_carrier_date'])
orders_df['order_delivered_customer_date'] = pd.to_datetime(orders_df['order_delivered_customer_date'])
orders_df['order_estimated_delivery_date'] = pd.to_datetime(orders_df['order_estimated_delivery_date'])

# Verify that there are no inaccurate values left
print(products_df.describe())
print(orders_df.info())


       product_name_lenght  product_description_lenght  product_photos_qty  \
count         31592.000000                31592.000000        31592.000000   
mean             48.429724                  767.406138            2.177452   
std              10.147889                  629.168854            1.716164   
min               5.000000                    4.000000            1.000000   
25%              42.000000                  341.000000            1.000000   
50%              51.000000                  600.000000            1.000000   
75%              57.000000                  951.000000            3.000000   
max              76.000000                 3992.000000           20.000000   

       product_weight_g  product_length_cm  product_height_cm  \
count      31592.000000       31592.000000       31592.000000   
mean        1827.888609          29.372519          15.911429   
std         3144.052398          14.553767          11.246668   
min            0.000000           7.0

##### Memperbaiki Tipe Data

In [132]:
# Correcting data types in the 'customers_df' DataFrame
customers_df['customer_zip_code_prefix'] = customers_df['customer_zip_code_prefix'].astype(str)

# Correcting data types in the 'geolocation_df' DataFrame
geolocation_df['geolocation_df_zip_code_prefix'] = geolocation_df['geolocation_df_zip_code_prefix'].astype(str)

# Correcting data types in the 'items_df' DataFrame
items_df['shipping_limit_date'] = pd.to_datetime(items_df['shipping_limit_date'])

# Correcting data types in the 'reviews_df' DataFrame
reviews_df['review_creation_date'] = pd.to_datetime(reviews_df['review_creation_date'])
reviews_df['review_answer_timestamp'] = pd.to_datetime(reviews_df['review_answer_timestamp'])

# Correcting data types in the 'sellers_df' DataFrame
sellers_df['seller_zip_code_prefix'] = sellers_df['seller_zip_code_prefix'].astype(str)

# Verify that the data types have been corrected
print(customers_df.dtypes)
print(geolocation_df.dtypes)
print(items_df.dtypes)
print(reviews_df.dtypes)
print(sellers_df.dtypes)

customer_id                 object
customer_unique_id          object
customer_zip_code_prefix    object
customer_city               object
customer_state              object
dtype: object
geolocation_df_zip_code_prefix     object
geolocation_df_lat                float64
geolocation_df_lng                float64
geolocation_df_city                object
geolocation_df_state               object
dtype: object
order_id                       object
order_item_id                   int64
product_id                     object
seller_id                      object
shipping_limit_date    datetime64[ns]
price                         float64
freight_value                 float64
dtype: object
review_id                          object
order_id                           object
review_score                        int64
review_comment_title               object
review_comment_message             object
review_creation_date       datetime64[ns]
review_answer_timestamp    datetime64[ns]
dtype: object

**Insight:**
1. **Penghapusan Duplikasi Data:**
    - Tidak ada duplikasi data yang ditemukan pada tabel `customers`, `geolocation`, `items`, `payments`, `reviews`, `orders`, `category`, `products`, dan `sellers` setelah proses penghapusan duplikasi dilakukan.

2. **Penanganan Nilai Hilang:**
    - Pada tabel `products`, nilai yang hilang pada kolom `product_category_name` diisi dengan 'unknown', sedangkan nilai yang hilang pada kolom `product_name_lenght`, `product_description_lenght`, `product_photos_qty`, `product_weight_g`, `product_length_cm`, `product_height_cm`, dan `product_width_cm` diisi dengan nilai rata-rata dari masing-masing kolom.
    - Pada tabel `orders`, nilai yang hilang pada kolom `order_approved_at`, `order_delivered_carrier_date`, dan `order_delivered_customer_date` diisi dengan metode forward fill.
    - Pada tabel `reviews`, nilai yang hilang pada kolom `review_comment_title` diisi dengan 'No Title' dan pada kolom `review_comment_message` diisi dengan 'No Comment'.
    - Setelah penanganan nilai hilang, tidak ada nilai yang hilang yang tersisa pada tabel `products`, `orders`, dan `reviews`.

3. **Penanganan Nilai Tidak Akurat:**
    - Pada tabel `products`, outlier pada kolom `product_weight_g`, `product_length_cm`, `product_height_cm`, dan `product_width_cm` dihapus dengan menggunakan kuantil 99%.
    - Pada tabel `orders`, tipe data pada kolom `order_purchase_timestamp`, `order_approved_at`, `order_delivered_carrier_date`, `order_delivered_customer_date`, dan `order_estimated_delivery_date` dikonversi menjadi tipe data datetime.

4. **Perbaikan Tipe Data:**
    - Pada tabel `customers`, kolom `customer_zip_code_prefix` dikonversi menjadi tipe data string.
    - Pada tabel `geolocation`, kolom `geolocation_zip_code_prefix` dikonversi menjadi tipe data string.
    - Pada tabel `items`, kolom `shipping_limit_date` dikonversi menjadi tipe data datetime.
    - Pada tabel `reviews`, kolom `review_creation_date` dan `review_answer_timestamp` dikonversi menjadi tipe data datetime.
    - Pada tabel `sellers`, kolom `seller_zip_code_prefix` dikonversi menjadi tipe data string.

## Exploratory Data Analysis (EDA)

### Explore `customers_df`

In [133]:
customers_df.sample(5)

customer_id                customer_unique_id  \
69110  b57a5ecb737d245afaf5bbcdd8e0375e  515d3550e900372e765ad518fccee139   
23317  de8e62a27572b1f6e9f0acd5c14a3026  47d19ed3e1569e33d47ed5712f16107c   
28277  3071ce9d57b701079af6747a0d89b13b  1427b47a463aa0edd349ca3a0e556729   
10012  5b7f46966e6e8992e9f9b0bb973ecc3d  3ca67cf83218dedd76e9f900261e3b27   
739    dfeb499afeccbe255d81f5900fb20879  88baa2a0a13e63c2e2b3ebc09595e4c8   

      customer_zip_code_prefix         customer_city customer_state  
69110                    45020  vitoria da conquista             BA  
23317                    90050          porto alegre             RS  
28277                    22755        rio de janeiro             RJ  
10012                    93290                esteio             RS  
739                      77440                gurupi             TO

In [150]:
customers_df.describe(include="all")

customer_id                customer_unique_id  \
count                              99441                             99441   
unique                             99441                             96096   
top     06b8999e2fba1a1fbc88172c00ba8bc7  8d50f5eadf50201ccdcedfb9e2ac8455   
freq                                   1                                17   

       customer_zip_code_prefix customer_city customer_state  
count                     99441         99441          99441  
unique                    14994          4119             27  
top                       22790     sao paulo             SP  
freq                        142         15540          41746

In [151]:
customers_df.customer_id.is_unique

True

In [152]:
customers_df.customer_id.duplicated

<bound method Series.duplicated of 0        06b8999e2fba1a1fbc88172c00ba8bc7
1        18955e83d337fd6b2def6b18a428ac77
2        4e7b3e00288586ebd08712fdd0374a03
3        b2b6027bc5c5109e529d4dc6358b12c3
4        4f2d8ab171c80ec8364f7c12e35b23ad
                       ...               
99436    17ddf5dd5d51696bb3d7c6291687be6f
99437    e7b71a9017aa05c9a7fd292d714858e8
99438    5e28dfe12db7fb50a4b2f691faecea5e
99439    56b18e2166679b8a959d72dd06da27f9
99440    274fa6071e5e17fe303b9748641082c8
Name: customer_id, Length: 99441, dtype: object>

In [134]:
customers_df.groupby(by="customer_city").customer_id.nunique().sort_values(ascending=False)

customer_city
sao paulo         15540
rio de janeiro     6882
belo horizonte     2773
brasilia           2131
curitiba           1521
                  ...  
ibiara                1
rio espera            1
rio dos indios        1
rio dos cedros        1
lagoao                1
Name: customer_id, Length: 4119, dtype: int64

In [135]:
customers_df.groupby(by="customer_state").customer_id.nunique().sort_values(
    ascending=False
)

customer_state
SP    41746
RJ    12852
MG    11635
RS     5466
PR     5045
SC     3637
BA     3380
DF     2140
ES     2033
GO     2020
PE     1652
CE     1336
PA      975
MT      907
MA      747
MS      715
PB      536
PI      495
RN      485
AL      413
SE      350
TO      280
RO      253
AM      148
AC       81
AP       68
RR       46
Name: customer_id, dtype: int64

### Explore `geolocation_df`

In [136]:
geolocation_df.sample(5)

geolocation_df_zip_code_prefix  geolocation_df_lat  geolocation_df_lng  \
950103                          91710          -30.075870          -51.200315   
270704                          12243          -23.205994          -45.899726   
212255                           8620          -23.622073          -46.296615   
322923                          13820          -22.682877          -46.980723   
750612                          63630           -5.449470          -39.717019   

        geolocation_df_city geolocation_df_state  
950103         porto alegre                   RS  
270704  sao jose dos campos                   SP  
212255               suzano                   SP  
322923           jaguariuna                   SP  
750612         pedra branca                   CE

In [138]:
geolocation_df.groupby("geolocation_df_zip_code_prefix").size().sort_values(ascending=False)

geolocation_df_zip_code_prefix
38400    779
35500    751
11680    727
11740    678
36400    627
        ... 
40024      1
70056      1
70055      1
70054      1
3840       1
Length: 19015, dtype: int64

In [139]:
geolocation_df[geolocation_df["geolocation_df_zip_code_prefix"] == 24230].head()

Empty DataFrame
Columns: [geolocation_df_zip_code_prefix, geolocation_df_lat, geolocation_df_lng, geolocation_df_city, geolocation_df_state]
Index: []

### Explore `items_df`

In [140]:
items_df.sample(5)

order_id  order_item_id  \
53313  794fe563dd7bc49998f96c29d3e5152a              1   
63115  90195b2be816e11d2c854069f663f7f0              1   
19483  2cb5d5cb410ad0b66df87bacd021c021              1   
86091  c391d14cd15fc8e3e71ef9dd58fdb6c3              1   
33371  4b95fbb9125e0baff8527de2d1d0042b              1   

                             product_id                         seller_id  \
53313  6189e02e0740fb983b93e8075e965290  7040e82f899a04d1b434b795a43b4617   
63115  cd3b8a153ef3be4fdbad9e011b7b35e2  ffeee66ac5d5a62fe688b9d26f83f534   
19483  0df2a7a1ee25a81b032358dbb083319b  63b464dbf392c7b80d12d932fa7cafed   
86091  9ab61676eb915c81c3c893adca21364c  dc4a0fc896dc34b0d5bfec8438291c80   
33371  74824accb4f777403366c8e6b4e32d9b  cce6ab8d1682639fe45ab70234f1665f   

      shipping_limit_date   price  freight_value  
53313 2017-11-30 17:53:27   29.90          11.85  
63115 2017-12-04 23:31:16  149.99          21.02  
19483 2017-08-04 18:35:22  169.00          14.79  
86091 2017-08-27 22:30:23   83.36          16.34  
33371 2017-09-06 07:05:08   40.00          16.60

### Explore `payments_df`

In [141]:
payments_df.sample(5)

order_id  payment_sequential payment_type  \
86546  8e6e7d37d25c7e66e66f8be5a34da04d                   1  credit_card   
77608  174c83fcd9eb7abaaba69d49ea92ba96                   1       boleto   
80493  b35ce82e5e535cc632c4441e01d6046a                   1   debit_card   
63704  fca2aaefc989d1a8eddc4cb03b160eca                   1       boleto   
77198  b3110047dd3a27ff4f565f9a7bf591de                   1  credit_card   

       payment_installments  payment_value  
86546                     2         171.77  
77608                     1         221.23  
80493                     1         117.88  
63704                     1          97.35  
77198                     1         111.20

In [142]:
payments_df.groupby(by="payment_type").order_id.nunique().sort_values(
    ascending=False
)

payment_type
credit_card    76505
boleto         19784
voucher         3866
debit_card      1528
not_defined        3
Name: order_id, dtype: int64

### Explore `reviews_df`

In [143]:
reviews_df.sample(5)

review_id                          order_id  \
47959  df2d6a305ec8fcbe55ab3462429dfe2b  29e40e2efb1493c943466c5e401d08ab   
74982  89fde8a748dd11177fd72d4ea5796654  0f39e640d202449c6fd3aebf9b09f1fd   
84398  5bd39ef8f3c59cbc9403c8b4527868ff  2343d4a98379202b11d3ec66c322ab57   
9887   46eb30f11c3451a4d21129e6f4d1e1a8  a013a9692edd311c7cd46675f807ab24   
58851  750765a8f6cdf43e90a057a8edb34327  a3bebc15aac7ca437c445ca941e5d152   

       review_score review_comment_title  \
47959             5             No Title   
74982             5             No Title   
84398             5             No Title   
9887              5             No Title   
58851             5      Super recomendo   

                                  review_comment_message review_creation_date  \
47959                                     Tudo perfeito            2017-04-04   
74982                                         No Comment           2018-08-21   
84398                                         No Comment           2018-04-06   
9887                                          No Comment           2017-06-30   
58851  A encomenda chegou antes da hora. Tudo veio co...           2018-08-22   

      review_answer_timestamp  
47959     2017-04-07 02:03:15  
74982     2018-08-21 22:46:31  
84398     2018-04-07 09:03:16  
9887      2017-06-30 23:06:09  
58851     2018-08-23 11:23:35

### Explore `payments_df` & `reviews_df`

In [155]:
payments_reviews_df = pd.merge(
    left=payments_df,
    right=reviews_df,
    how="left",
    left_on="order_id",
    right_on="order_id",
)
payments_reviews_df.head()

order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   

   payment_installments  payment_value                         review_id  \
0                     8          99.33  4ce650e117a5b292bf5ea980a61eb61e   
1                     1          24.39  ad2036caca61484fc435d7e673c020f2   
2                     1          65.71  bcddfa0f7d2e54cfc74bae4f01b84d92   
3                     8         107.78  c37a0e51e5df82bdb55d38ff0ffb6af3   
4                     8         107.78  5bc7680443b4ef765fec11a8201eeb0c   

   review_score review_comment_title  \
0           1.0        Não recomendo   
1           5.0             No Title   
2           5.0             No Title   
3           5.0             No Title   
4           5.0             No Title   

                              review_comment_message review_creation_date  \
0  Produto com qualidade ruim, amassado e descasc...           2018-05-10   
1                                         No Comment           2018-06-30   
2                                           Aprovada           2017-12-19   
3                                         No Comment           2017-12-21   
4                                         No Comment           2017-12-21   

  review_answer_timestamp  
0     2018-05-15 23:38:00  
1     2018-07-03 11:23:16  
2     2017-12-22 17:55:30  
3     2017-12-21 21:42:31  
4     2017-12-22 01:10:39

In [156]:
payments_reviews_df.sort_values(by="payment_value", ascending=False)

order_id  payment_sequential payment_type  \
52415  03caa2c082116e1d31e67e9ae3700499                   1  credit_card   
34565  736e1922ae60d0d6a89247b851902527                   1       boleto   
41660  0812eb902a67711a1cb742b3cdaa65ae                   1  credit_card   
49871  fefacc66af859508bf1a7934eab1e97f                   1       boleto   
86018  f5136e38d1a14a4dbd87dff67da82701                   1       boleto   
...                                 ...                 ...          ...   
78326  fa65dad1b0e818e3ccc5cb0e39231352                  13      voucher   
51581  4637ca194b6387e2d538dc89b124b0ee                   1  not_defined   
37030  fa65dad1b0e818e3ccc5cb0e39231352                  14      voucher   
94953  c8c528189310eaa44a745b8d9d26908b                   1  not_defined   
20031  8bcbe01d44d147f901cd3192671144db                   4      voucher   

       payment_installments  payment_value                         review_id  \
52415                     1       13664.08  211fa81a22870cc634265ea85b751e40   
34565                     1        7274.88  6b9a600f1bb824da4ca38f65baf42526   
41660                     8        6929.31  e2fbd9f0ef7b724c7ecb351601a857e7   
49871                     1        6922.21                               NaN   
86018                     1        6726.66  1b73e6c7cd135008ebeb18404b776ddc   
...                     ...            ...                               ...   
78326                     1           0.00  f28281373ab8815bafafe371218f02ce   
51581                     1           0.00  9efef4d70190c1bdb9b12872cc8144fc   
37030                     1           0.00  f28281373ab8815bafafe371218f02ce   
94953                     1           0.00  b54d02f7f1520b3995f84b77ab0dacb3   
20031                     1           0.00  3d42038fd2a7d35a88b93cb3e51796a3   

       review_score       review_comment_title  \
52415           1.0                   No Title   
34565           1.0         Entrega incompleta   
41660           5.0                   No Title   
49871           NaN                        NaN   
86018           5.0                   No Title   
...             ...                        ...   
78326           1.0                   No Title   
51581           3.0           Produto Avariado   
37030           1.0                   No Title   
94953           1.0  Péssima meu produtdefeito   
20031           4.0                   No Title   

                                  review_comment_message review_creation_date  \
52415  Peço por favor verificar a entrega do pedido, ...           2017-10-18   
34565  Comprei 04 unidades, e foram entregues apenas ...           2018-07-27   
41660  Mesmo a máquina ter vindo errada, já se pronti...           2017-03-04   
49871                                                NaN                  NaT   
86018                                         No Comment           2017-06-06   
...                                                  ...                  ...   
78326  A mercadoria não foi entregue. Entrara em cont...           2017-05-21   
51581  Recebi o Diario do Bebê c um pequeno rasgo na ...           2018-08-17   
37030  A mercadoria não foi entregue. Entrara em cont...           2017-05-21   
94953                                         No Comment           2018-08-24   
20031                                         No Comment           2018-02-03   

      review_answer_timestamp  
52415     2017-10-19 13:10:39  
34565     2018-07-28 02:36:10  
41660     2017-03-04 21:40:41  
49871                     NaT  
86018     2017-06-07 19:35:23  
...                       ...  
78326     2017-05-24 16:21:27  
51581     2018-08-18 15:32:40  
37030     2017-05-24 16:21:27  
94953     2018-08-25 02:24:54  
20031     2018-02-11 17:34:29  

[104477 rows x 11 columns]

In [157]:
payments_reviews_df.groupby(by="payment_type").agg(
    {"order_id": "nunique", "payment_value": ["min", "max"]}
)

order_id payment_value          
              nunique           min       max
payment_type                                 
boleto          19784         11.62   7274.88
credit_card     76505          0.01  13664.08
debit_card       1528         13.38   4445.50
not_defined         3          0.00      0.00
voucher          3866          0.00   3184.34

### Explore `orders_df`

In [144]:
orders_df.sample(5)

order_id                       customer_id  \
8583   b29f6f52b330f04b5f1f4fb369b32cd8  8c0bab5a25230a9d6500f31081041e21   
21177  d24cf30658f279bb1e8677169e43e44a  933da2a653210f730747c28a49d54d8e   
63987  37ec91b7248a28948cfedf05c2a82b15  0443dd75d52f59a5b0061b149534737e   
35899  3cd4d6461c966c84e6547503dc04df2a  26d81167c70403d2026325910c1dffbd   
13969  6192897f85cb2aff6dd1fca56fddb45e  f77559535865a660bd328db79a177749   

      order_status order_purchase_timestamp   order_approved_at  \
8583     delivered      2018-07-13 12:01:24 2018-07-13 12:10:20   
21177    delivered      2018-07-27 13:12:48 2018-07-27 13:25:22   
63987    delivered      2017-11-29 14:08:40 2017-11-29 14:29:50   
35899    delivered      2017-11-24 12:27:36 2017-11-24 14:34:25   
13969    delivered      2018-07-18 13:34:22 2018-07-18 13:50:12   

      order_delivered_carrier_date order_delivered_customer_date  \
8583           2018-07-13 14:52:00           2018-07-23 21:59:33   
21177          2018-07-30 13:51:00           2018-08-04 12:18:56   
63987          2017-11-30 21:09:46           2017-12-01 18:54:35   
35899          2017-11-28 21:52:18           2017-12-04 19:58:53   
13969          2018-07-18 11:33:00           2018-07-26 15:02:53   

      order_estimated_delivery_date  
8583                     2018-08-01  
21177                    2018-08-13  
63987                    2017-12-13  
35899                    2017-12-14  
13969                    2018-08-03

In [153]:
orders_df.describe(include="all")

order_id                       customer_id  \
count                              99441                             99441   
unique                             99441                             99441   
top     e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
freq                                   1                                 1   
mean                                 NaN                               NaN   
min                                  NaN                               NaN   
25%                                  NaN                               NaN   
50%                                  NaN                               NaN   
75%                                  NaN                               NaN   
max                                  NaN                               NaN   

       order_status       order_purchase_timestamp  \
count         99441                          99441   
unique            8                            NaN   
top       delivered                            NaN   
freq          96478                            NaN   
mean            NaN  2017-12-31 08:43:12.776581120   
min             NaN            2016-09-04 21:15:19   
25%             NaN            2017-09-12 14:46:19   
50%             NaN            2018-01-18 23:04:36   
75%             NaN            2018-05-04 15:42:16   
max             NaN            2018-10-17 17:30:18   

                    order_approved_at   order_delivered_carrier_date  \
count                           99441                          99441   
unique                            NaN                            NaN   
top                               NaN                            NaN   
freq                              NaN                            NaN   
mean    2017-12-31 20:23:34.460584704  2018-01-05 01:00:33.728884480   
min               2016-09-15 12:16:38            2016-10-08 10:34:01   
25%               2017-09-13 03:15:36            2017-09-18 13:50:17   
50%               2018-01-19 13:35:58            2018-01-24 17:49:18   
75%               2018-05-04 21:35:13            2018-05-08 13:51:00   
max               2018-09-03 17:40:06            2018-09-11 19:48:28   

        order_delivered_customer_date  order_estimated_delivery_date  
count                           99441                          99441  
unique                            NaN                            NaN  
top                               NaN                            NaN  
freq                              NaN                            NaN  
mean    2018-01-14 15:01:35.991602944  2018-01-24 03:08:37.730111232  
min               2016-10-11 13:46:32            2016-09-30 00:00:00  
25%               2017-09-26 00:48:43            2017-10-03 00:00:00  
50%               2018-02-02 20:08:54            2018-02-15 00:00:00  
75%               2018-05-15 23:21:53            2018-05-25 00:00:00  
max               2018-10-17 13:22:46            2018-11-12 00:00:00

In [146]:
orders_df["order_status"].describe()

count         99441
unique            8
top       delivered
freq          96478
Name: order_status, dtype: object

### Explore `customers_df` & `orders_df`

In [154]:
customers_orders_df = pd.merge(
    left=customers_df,
    right=orders_df,
    how="left",
    left_on="customer_id",
    right_on="customer_id",
)
customers_orders_df.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

  customer_zip_code_prefix          customer_city customer_state  \
0                    14409                 franca             SP   
1                     9790  sao bernardo do campo             SP   
2                     1151              sao paulo             SP   
3                     8775        mogi das cruzes             SP   
4                    13056               campinas             SP   

                           order_id order_status order_purchase_timestamp  \
0  00e7ee1b050b8499577073aeb2a297a1    delivered      2017-05-16 15:05:35   
1  29150127e6685892b6eab3eec79f59c7    delivered      2018-01-12 20:48:24   
2  b2059ed67ce144a36e2aa97d2c9e9ad2    delivered      2018-05-19 16:07:45   
3  951670f92359f4fe4a63112aa7306eba    delivered      2018-03-13 16:06:38   
4  6b7d50bd145f6fc7f33cebabd7e49d0f    delivered      2018-07-29 09:51:30   

    order_approved_at order_delivered_carrier_date  \
0 2017-05-16 15:22:12          2017-05-23 10:47:57   
1 2018-01-12 20:58:32          2018-01-15 17:14:59   
2 2018-05-20 16:19:10          2018-06-11 14:31:00   
3 2018-03-13 17:29:19          2018-03-27 23:22:42   
4 2018-07-29 10:10:09          2018-07-30 15:16:00   

  order_delivered_customer_date order_estimated_delivery_date  
0           2017-05-25 10:35:35                    2017-06-05  
1           2018-01-29 12:41:19                    2018-02-06  
2           2018-06-14 17:58:51                    2018-06-13  
3           2018-03-28 16:04:25                    2018-04-10  
4           2018-08-09 20:55:48                    2018-08-15

### Explore `category_df`

In [147]:
category_df.sample(5)

product_category_df_name product_category_df_name_english
4                    moveis_decoracao                  furniture_decor
30                  moveis_escritorio                 office_furniture
62                    casa_conforto_2                   home_comfort_2
43  construcao_ferramentas_iluminacao        construction_tools_lights
1              informatica_acessorios            computers_accessories

### Explore `products_df`

In [148]:
products_df.sample(5)

product_id product_category_df_name  \
20967  a58b5955d37aea2de1ff80d5cba4f8ba         moveis_decoracao   
10954  f49f59559a40c88ed4beeacac8b25e71  fashion_roupa_masculina   
22872  05a4cea3c80d33b04145946e6c5e0fa9       ferramentas_jardim   
5058   d73068913476e74d4f35a6f2d5399ae9         malas_acessorios   
18272  1cbae91a0159b540e3d3a43193d3e970       relogios_presentes   

       product_name_lenght  product_description_lenght  product_photos_qty  \
20967                 41.0                       900.0                 1.0   
10954                 28.0                       364.0                 2.0   
22872                 50.0                       536.0                 1.0   
5058                  40.0                      1070.0                 1.0   
18272                 56.0                       555.0                 1.0   

       product_weight_g  product_length_cm  product_height_cm  \
20967             500.0               27.0               35.0   
10954             250.0               16.0               16.0   
22872             850.0               22.0                7.0   
5058              650.0               90.0                5.0   
18272             150.0               20.0                5.0   

       product_width_cm  
20967              22.0  
10954              16.0  
22872              17.0  
5058               11.0  
18272              15.0

### Explore `products_df` & `category_df`

In [168]:
products_category_df = pd.merge(
    left=products_df,
    right=category_df,
    how="left",
    left_on="product_category_df_name",
    right_on="product_category_df_name",
)
products_category_df.head()

product_id product_category_df_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5               perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                    artes   
2  96bd76ec8810374ed1b65e291975717f            esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                    bebes   
4  9dc1a7de274444849c219cff195d0b71    utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  \
0             225.0               16.0               10.0              14.0   
1            1000.0               30.0               18.0              20.0   
2             154.0               18.0                9.0              15.0   
3             371.0               26.0                4.0              26.0   
4             625.0               20.0               17.0              13.0   

  product_category_df_name_english  
0                        perfumery  
1                              art  
2                   sports_leisure  
3                             baby  
4                       housewares

In [169]:
products_category_df.groupby(
    by="product_category_df_name"
).product_id.nunique().sort_values(ascending=False).head(10)

product_category_df_name
cama_mesa_banho           2966
esporte_lazer             2737
beleza_saude              2390
moveis_decoracao          2363
utilidades_domesticas     2216
automotivo                1813
informatica_acessorios    1633
brinquedos                1383
relogios_presentes        1327
telefonia                 1134
Name: product_id, dtype: int64

In [170]:
products_category_df.groupby(
    by="product_category_df_name_english"
).product_id.nunique().sort_values(ascending=False).head(10)

product_category_df_name_english
bed_bath_table           2966
sports_leisure           2737
health_beauty            2390
furniture_decor          2363
housewares               2216
auto                     1813
computers_accessories    1633
toys                     1383
watches_gifts            1327
telephony                1134
Name: product_id, dtype: int64

### Explore `sellers_df`

In [149]:
sellers_df.sample(5)

seller_id seller_zip_code_prefix  \
996   06bb3a2fe5e7b7a845b13e8fb91bd944                  31844   
498   e4f121bf6ef8b9a1d4d3e65dd0473fab                  15085   
2424  5d505d2298ad549e4e4e835562fa0c1e                  16200   
1986  e59e489e2ea74bd409b798b5daf49de3                  87060   
285   a3a38f4affed601eb87a97788c949667                  89204   

                seller_city seller_state  
996          belo horizonte           MG  
498   sao jose do rio preto           SP  
2424                birigui           SP  
1986                maringa           PR  
285               joinville           SC

### Explore `items_df` & `sellers_df`

In [158]:
item_seller_df = pd.merge(
    left=items_df,
    right=sellers_df,
    how="left",
    left_on="seller_id",
    right_on="seller_id",
)
item_seller_df.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

  shipping_limit_date   price  freight_value seller_zip_code_prefix  \
0 2017-09-19 09:45:35   58.90          13.29                  27277   
1 2017-05-03 11:05:13  239.90          19.93                   3471   
2 2018-01-18 14:48:30  199.00          17.87                  37564   
3 2018-08-15 10:10:18   12.99          12.79                  14403   
4 2017-02-13 13:57:51  199.90          18.14                  87900   

     seller_city seller_state  
0  volta redonda           SP  
1      sao paulo           SP  
2  borda da mata           MG  
3         franca           SP  
4         loanda           PR

In [159]:
item_seller_df.groupby(by="seller_city").seller_id.nunique().sort_values(
    ascending=False
).head(10)

seller_city
sao paulo         694
curitiba          127
rio de janeiro     96
belo horizonte     68
ribeirao preto     52
guarulhos          50
ibitinga           49
santo andre        45
campinas           41
maringa            40
Name: seller_id, dtype: int64

In [160]:
item_seller_df.groupby(by="seller_state").seller_id.nunique().sort_values(
    ascending=False
).head(10)

seller_state
SP    1849
PR     349
MG     244
SC     190
RJ     171
RS     129
GO      40
DF      30
ES      23
BA      19
Name: seller_id, dtype: int64

### Menggabungkan Data

In [ ]:
# Merge orders_df with customers_df
orders_df_customers_df = pd.merge(orders_df, customers_df, on='customer_id', how='inner')

# Merge orders_df_customers_df with payments_df
orders_df_customers_df_payments_df = pd.merge(orders_df_customers_df, payments_df, on='order_id', how='inner')

# Merge orders_df_customers_df_payments_df with reviews_df
orders_df_customers_df_payments_df_reviews_df = pd.merge(orders_df_customers_df_payments_df, reviews_df, on='order_id', how='inner')

# Merge orders_df_customers_df_payments_df_reviews_df with items_df
orders_df_customers_df_payments_df_reviews_df_items_df = pd.merge(orders_df_customers_df_payments_df_reviews_df, items_df, on='order_id', how='inner')

# Merge orders_df_customers_df_payments_df_reviews_df_items_df with products_df
orders_df_customers_df_payments_df_reviews_df_items_df_products_df = pd.merge(orders_df_customers_df_payments_df_reviews_df_items_df, products_df, on='product_id', how='inner')

# Merge orders_df_customers_df_payments_df_reviews_df_items_df_products_df with category_df
orders_df_customers_df_payments_df_reviews_df_items_df_products_df_category_df = pd.merge(orders_df_customers_df_payments_df_reviews_df_items_df_products_df, category_df, on='product_category_df_name', how='inner')

# Merge orders_df_customers_df_payments_df_reviews_df_items_df_products_df_category_df with sellers_df
final_df = pd.merge(orders_df_customers_df_payments_df_reviews_df_items_df_products_df_category_df, sellers_df, on='seller_id', how='inner')

# Display the first few rows of the final dataframe
final_df.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
2  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
3  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
4  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   

  order_status order_purchase_timestamp   order_approved_at  \
0    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
2    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
3    delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
4    delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2017-10-04 19:55:00           2017-10-10 21:25:13   
2          2017-10-04 19:55:00           2017-10-10 21:25:13   
3          2018-07-26 14:31:00           2018-08-07 15:27:45   
4          2018-08-08 13:50:00           2018-08-17 18:06:29   

  order_estimated_delivery_date                customer_unique_id  \
0                    2017-10-18  7c396fd4830fd04220f754e42b4e5bff   
1                    2017-10-18  7c396fd4830fd04220f754e42b4e5bff   
2                    2017-10-18  7c396fd4830fd04220f754e42b4e5bff   
3                    2018-08-13  af07308b275d755c9edb36a90c618231   
4                    2018-09-04  3a653a41f6f9fc3d2a113cf8398680e8   

  customer_zip_code_prefix  ... product_description_lenght product_photos_qty  \
0                     3149  ...                      268.0                4.0   
1                     3149  ...                      268.0                4.0   
2                     3149  ...                      268.0                4.0   
3                    47813  ...                      178.0                1.0   
4                    75265  ...                      232.0                1.0   

   product_weight_g product_length_cm  product_height_cm  product_width_cm  \
0             500.0              19.0                8.0              13.0   
1             500.0              19.0                8.0              13.0   
2             500.0              19.0                8.0              13.0   
3             400.0              19.0               13.0              19.0   
4             420.0              24.0               19.0              21.0   

  product_category_name_english  seller_zip_code_prefix     seller_city  \
0                    housewares                    9350            maua   
1                    housewares                    9350            maua   
2                    housewares                    9350            maua   
3                     perfumery                   31570  belo horizonte   
4                          auto                   14840         guariba   

  seller_state  
0           SP  
1           SP  
2           SP  
3           SP  
4           SP  

[5 rows x 40 columns]

In [ ]:
# Export the final dataframe to a CSV file
final_df.to_csv("./dataset/final_dataframe.csv", index=False)

**Insight:**
- xxx
- xxx

## Visualization & Explanatory Analysis

In [ ]:
# Distribution of customers_df by city
customer_city_distribution = customers_df['customer_city'].value_counts().head(10)
print("Top 10 Cities with Most Customers:")
print(customer_city_distribution)

# Distribution of customers_df by state
customer_state_distribution = customers_df['customer_state'].value_counts()
print("\nDistribution of Customers by State:")
print(customer_state_distribution)

# Number of unique customers_df
unique_customers_df = customers_df['customer_unique_id'].nunique()
print("\nNumber of Unique Customers:", unique_customers_df)

# Plotting the distribution of customers_df by city
plt.figure(figsize=(12, 6))
customer_city_distribution.plot(kind='bar')
plt.title('Top 10 Cities with Most Customers')
plt.xlabel('City')
plt.ylabel('Number of Customers')
plt.xticks(rotation=45)
plt.show()

# Plotting the distribution of customers_df by state
plt.figure(figsize=(12, 6))
customer_state_distribution.plot(kind='bar')
plt.title('Distribution of Customers by State')
plt.xlabel('State')
plt.ylabel('Number of Customers')
plt.xticks(rotation=45)
plt.show()

### Pertanyaan 1:

### Pertanyaan 2:

**Insight:**
- xxx
- xxx

## Analisis Lanjutan (Opsional)

## Conclusion

- Conclution pertanyaan 1
- Conclution pertanyaan 2